### PR TITLE
[WEB-696] CTA Dialog on Facebook Login Error (FB Deprecation)

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

As a User who does not have a Kickstarter password and attempts to log in via Facebook, but gets an error either from Facebook, we should present a CTA dialog on the login screen with options to set a new password or use normal email password login

See [the designs](https://www.figma.com/file/lmJlSlEk5Wp7BfxCVtKplU/Deprecating-FB-Authentication?node-id=293%3A810&t=gjl1jEnxGVURvLKA-0) for details on what this should look like.

Since we're still looking into the issue we're seeing around snapshot testing on M1 macbooks, I've created [a ticket](https://kickstarter.atlassian.net/browse/NTV-638) to get that part of this PR in once, Mubarak is back

# 🤔 Why

In the past FB has turned off their FB Login API without warning keeping kickstarters from logging in. 
In an effort to deprecate FB Login altogether, this covers that scenario. Both informing the kickstarter that we need them to set a password or log in through our normal login flow.

This will allow us to ensure that everyone is more aware of the move away from FB Login as well as give everyone a chance to set a password on their accounts if they don't already have one.

# 🛠 How

- Created a new `ResetYourFacebookPassword` screen. Allows a kickstarter to get an email to set a password on their account. This is essentially the normal reset password flow.
- Created a new CTA popup with 2 options. Either navigate to normal log in _or_ to the new screen.
- Using the `showFacebookErrorAlert` listener in `LoginToutViewController`, we can:
    1. Check to make sure that Facebook Login Deprecation flag is enabled
    2. Present this new CTA at the right time

# 👀 See

Trello, screenshots, external resources?

Simulator Screen Recording - iPhone 13 Pro - 2022-12-07 at 12 12 07

# ✅ Acceptance criteria

- [x] If an error occurs when attempting to login with FB, and the FF is on, this new CTA shows
- [x] The actions on the new CTA navigate to the correct screens
- [x] The submit button on `ResetYourFacebookPassword` screen sends an email to the email, if valid, allowing you to set a password on your account.

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by msadoon — COMMENTED** [review]
Hey Scott, generally works well, code is relatively clean.
I did a deeper review of this and found some improvements to make.

I didn't write it as a comment but can you also remove all the recorded screenshots here. I've made a note to record these in the fix screenshots [pr](https://github.com/kickstarter/ios-oss/pull/1757) so they are platform agnostic.

There's something else too which I know can be a chore, but would improve the accessibility and usability of the app for larger font sizes. It would also be more consistent with `SignUpViewController` and other form fields. 

The two things are:
1. Ensure the scrollview is not hidden by the keyboard when its' up. Context for how to achieve this is `SignupViewController` `animateTextViewConstraint` function.
2. Also the more complicated thing but does make sense with both `ResetYourFacebookPasswordViewController` and `SetYourPasswordViewController` is to allow each textfield to move to the next and then hook up the done button to the api call on the submit button. Examples of how to do this are in `SignupViewController` signals like `emailTextFieldReturn`, `nameTextFieldReturn` and `signupButtonPressed`.

https://user-images.githubusercontent.com/4282741/210671326-ce349a32-e47a-4cee-9670-3d17692e28fc.mp4

https://user-images.githubusercontent.com/4282741/210671329-e0b82ba8-ee29-4e06-b05a-1e78d309ad6b.mp4

Those last two are big asks, so if you want me to make the changes let me know.

Maybe you can write all this up in a separate "improvements" pr?

File: Library/ViewModels/ResetYourFacebookPasswordViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    let submitFormEvent = self.setPasswordButtonPressedProperty.signal
+
+    let submitAction = formIsValid
+      .takeWhen(submitFormEvent)
+      .filter(isTrue)
+      .ignoreValues()
+
+    let setPasswordEvent = self.emailTextFieldProperty.signal.skipNil()
+      .takeWhen(self.setPasswordButtonPressedProperty.signal)
```
[the remark this answers is not in the document]
Comment (scottkicks) [reply Library/ViewModels/ResetYourFacebookPasswordViewModel.swift:54]: yeah nice

File: Library/ViewModels/ResetYourFacebookPasswordViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    self.vm.inputs.emailTextFieldFieldDidChange("bad")
+
+    self.setPasswordButtonIsEnabled.assertValues([false])
+
+    self.vm.inputs.emailTextFieldFieldDidChange("gina@kickstarter.com")
+
+    self.setPasswordButtonIsEnabled.assertValues([false, true])
+  }
+
+  func testsetPasswordButtonIsEnabled_WithInitialValue() {
```
[the remark this answers is not in the document]
Comment (scottkicks) [reply Library/ViewModels/ResetYourFacebookPasswordViewModelTests.swift:54]: good point

File: Kickstarter-iOS/Features/ResetYourFacebookPassword/Controller/ResetYourFacebookPasswordViewController.swift [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+import Prelude
+import ReactiveSwift
+import UIKit
+
+public final class ResetYourFacebookPasswordViewController: UIViewController {
+  // MARK: - Properties
+
+  private lazy var contextLabel = { UILabel(frame: .zero) }()
+  private lazy var emailLabel: UILabel = { UILabel(frame: .zero) }()
+  private lazy var emailTextField: UITextField = { UITextField(frame: .zero) |> \.tag .~ 0 }()
```
[the remark this answers is not in the document]
Comment (scottkicks) [reply Kickstarter-iOS/Features/ResetYourFacebookPassword/Controller/ResetYourFacebookPasswordViewController.swift:12]: oops

File: Library/ViewModels/ResetYourFacebookPasswordViewModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+      self.setPasswordFailure.assertValues(
+        [Strings.forgot_password_error()],
+        "Error alert is shown on bad request"
+      )
+    }
+  }
+
+  func testResetFail_WithNon404Error() {
+    let error = ErrorEnvelope(
+      errorMessages: ["Zoinks!"],
```
Comment (msadoon) [inline Library/ViewModels/ResetYourFacebookPasswordViewModelTests.swift:113]: haha nice

**Pull request merged**